// @codemirror/next/lang-css/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var lezerCss = require('lezer-css');
var syntax = require('../../syntax');
var highlight = require('../../highlight');

/// A syntax provider based on the [Lezer CSS
/// parser](https://github.com/lezer-parser/css), extended with
/// highlighting and indentation information.
const cssSyntax = new syntax.LezerSyntax(lezerCss.parser.withProps(syntax.indentNodeProp.add({
    Declaration: syntax.continuedIndent()
}), syntax.foldNodeProp.add({
    Block(subtree) { return { from: subtree.start + 1, to: subtree.end - 1 }; }
}), highlight.styleTags({
    "import charset namespace keyframes": "keyword definition",
    "media supports": "keyword control",
    "from to": "keyword",
    NamespaceName: "namespace",
    KeyframeName: "labelName",
    TagName: "typeName",
    ClassName: "className",
    PseudoClassName: "className constant",
    not: "operatorKeyword",
    IdName: "labelName",
    AttributeName: "propertyName",
    NumberLiteral: "number",
    PropertyName: "propertyName",
    KeywordQuery: "keyword",
    FeatureName: "propertyName",
    UnaryQueryOp: "operatorKeyword",
    callee: "keyword",
    ValueName: "atom",
    CallTag: "atom",
    Callee: "variableName",
    Unit: "unit",
    "UniversalSelector NestingSelector": "operator definition",
    AtKeyword: "keyword",
    MatchOp: "compareOperator",
    "ChildOp SiblingOp, LogicOp": "logicOperator",
    BinOp: "arithmeticOperator",
    Important: "modifier",
    Comment: "blockComment",
    ParenthesizedContent: "name#2",
    ColorLiteral: "color",
    StringLiteral: "string",
    ":": "punctuation definition",
    "PseudoOp #": "derefOperator",
    "; ,": "separator",
    "( )": "paren",
    "[ ]": "squareBracket",
    "{ }": "brace"
})));
/// Returns an extension that installs the CSS syntax provider.
function css() { return cssSyntax.extension; }

exports.css = css;
exports.cssSyntax = cssSyntax;
