// @codemirror/next/matchbrackets/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var state = require('../../state');
var view = require('../../view');
var syntax = require('../../syntax');

const baseTheme = view.EditorView.baseTheme({
    matchingBracket: { color: "#0b0" },
    nonmatchingBracket: { color: "#a22" }
});
const DefaultScanDist = 10000, DefaultBrackets = "()[]{}";
const bracketMatchingConfig = state.Facet.define({
    combine(configs) {
        return state.combineConfig(configs, {
            afterCursor: true,
            brackets: DefaultBrackets,
            maxScanDistance: DefaultScanDist
        });
    }
});
const matchingMark = view.Decoration.mark({ class: view.themeClass("matchingBracket") }), nonmatchingMark = view.Decoration.mark({ class: view.themeClass("nonmatchingBracket") });
const bracketMatchingState = state.StateField.define({
    create() { return view.Decoration.none; },
    update(deco, tr, state) {
        if (!tr.docChanged && !tr.selectionSet)
            return deco;
        let decorations = [];
        let config = state.facet(bracketMatchingConfig);
        for (let range of state.selection.ranges) {
            if (!range.empty)
                continue;
            let match = matchBrackets(state, range.head, -1, config)
                || (range.head > 0 && matchBrackets(state, range.head - 1, 1, config))
                || (config.afterCursor &&
                    (matchBrackets(state, range.head, 1, config) ||
                        (range.head < state.doc.length && matchBrackets(state, range.head + 1, -1, config))));
            if (!match)
                continue;
            let mark = match.matched ? matchingMark : nonmatchingMark;
            decorations.push(mark.range(match.start.from, match.start.to));
            if (match.end)
                decorations.push(mark.range(match.end.from, match.end.to));
        }
        return view.Decoration.set(decorations, true);
    }
}).provide(view.EditorView.decorations);
const bracketMatchingUnique = [
    bracketMatchingState,
    baseTheme
];
/// Create an extension that enables bracket matching. Whenever the
/// cursor is next to a bracket, that bracket and the one it matches
/// are highlighted. Or, when no matching bracket is found, another
/// highlighting style is used to indicate this.
function bracketMatching(config = {}) {
    return [bracketMatchingConfig.of(config), bracketMatchingUnique];
}
function matchingNodes(node, dir, brackets) {
    let byProp = node.prop(dir < 0 ? syntax.closeNodeProp : syntax.openNodeProp);
    if (byProp)
        return byProp;
    if (node.name.length == 1) {
        let index = brackets.indexOf(node.name);
        if (index > -1 && index % 2 == (dir < 0 ? 1 : 0))
            return [brackets[index + dir]];
    }
    return null;
}
/// Find the matching bracket for the token at `pos`, scanning
/// direction `dir`. Only the `brackets` and `maxScanDistance`
/// properties are used from `config`, if given. Returns null if no
/// bracket was found at `pos`, or a match result otherwise.
function matchBrackets(state, pos, dir, config = {}) {
    let maxScanDistance = config.maxScanDistance || DefaultScanDist, brackets = config.brackets || DefaultBrackets;
    let tree = state.tree, sub = tree.resolve(pos, dir), matches;
    if (matches = matchingNodes(sub.type, dir, brackets))
        return matchMarkedBrackets(state, pos, dir, sub, matches, brackets);
    else
        return matchPlainBrackets(state, pos, dir, tree, sub.type, maxScanDistance, brackets);
}
function matchMarkedBrackets(_state, _pos, dir, token, matching, brackets) {
    let parent = token.parent, firstToken = { from: token.start, to: token.end };
    let depth = 0;
    return (parent && parent.iterate({
        from: dir < 0 ? token.start : token.end,
        to: dir < 0 ? parent.start : parent.end,
        enter(type, from, to) {
            if (dir < 0 ? to > token.start : from < token.end)
                return undefined;
            if (depth == 0 && matching.indexOf(type.name) > -1) {
                return { start: firstToken, end: { from, to }, matched: true };
            }
            else if (matchingNodes(type, dir, brackets)) {
                depth++;
            }
            else if (matchingNodes(type, -dir, brackets)) {
                depth--;
                if (depth == 0)
                    return { start: firstToken, end: { from, to }, matched: false };
            }
            return false;
        }
    })) || { start: firstToken, matched: false };
}
function matchPlainBrackets(state, pos, dir, tree, tokenType, maxScanDistance, brackets) {
    let startCh = dir < 0 ? state.doc.slice(pos - 1, pos) : state.doc.slice(pos, pos + 1);
    let bracket = brackets.indexOf(startCh);
    if (bracket < 0 || (bracket % 2 == 0) != (dir > 0))
        return null;
    let startToken = { from: dir < 0 ? pos - 1 : pos, to: dir > 0 ? pos + 1 : pos };
    let iter = state.doc.iterRange(pos, dir > 0 ? state.doc.length : 0), depth = 0;
    for (let distance = 0; !(iter.next()).done && distance <= maxScanDistance;) {
        let text = iter.value;
        if (dir < 0)
            distance += text.length;
        let basePos = pos + distance * dir;
        for (let pos = dir > 0 ? 0 : text.length - 1, end = dir > 0 ? text.length : -1; pos != end; pos += dir) {
            let found = brackets.indexOf(text[pos]);
            if (found < 0 || tree.resolve(basePos + pos, 1).type != tokenType)
                continue;
            if ((found % 2 == 0) == (dir > 0)) {
                depth++;
            }
            else if (depth == 1) { // Closing
                return { start: startToken, end: { from: basePos + pos, to: basePos + pos + 1 }, matched: (found >> 1) == (bracket >> 1) };
            }
            else {
                depth--;
            }
        }
        if (dir > 0)
            distance += text.length;
    }
    return iter.done ? { start: startToken, matched: false } : null;
}

exports.bracketMatching = bracketMatching;
exports.matchBrackets = matchBrackets;
